### PR TITLE
Fix Barreiss algorithm

### PR DESCRIPTION
**Pull request by oscardssmith**
This has a number of changes.
The first is that we were using ÷ instead of exact_div in bareiss_update_virtual_colswap_mtk which led to undetected overflows. We also had a number of bugs that arise when using BigInt precsion in bareiss (e.g. checking `!==0` vs `!=0`. This PR makes it so we first try to bareiss with the regular Int64 matrix, but if that gets an overflow, we fallback to the BigInt version.
@YingboMa I'm not 100% convinced that the `exact_div` is a sufficient check. Is it obvious that it always suffices to check the remainder or should we instead be checking for overflow on the multiplications?

### CODE REVIEW

File: src/systems/alias_elimination.jl [hunk 1]
```diff
@@ -996,7 +1005,7 @@ function locally_structure_simplify!(adj_row, pivot_var, ag)
         alias = get(ag, var, nothing)
         if alias === nothing
             nirreducible += 1
-            alias_candidate = val => var
```
Comment (YingboMa) [inline src/systems/alias_elimination.jl:999]: How do we know `val` can always be stored in `Int`?

File: src/systems/sparsematrixclil.jl [hunk 2]
```diff
@@ -29,11 +29,20 @@ function Base.copy(S::SparseMatrixCLIL{T, Ti}) where {T, Ti}
                      map(copy, S.row_vals))
 end
 function swaprows!(S::SparseMatrixCLIL, i, j)
+    i == j && return
     swap!(S.nzrows, i, j)
     swap!(S.row_cols, i, j)
     swap!(S.row_vals, i, j)
 end
 
+function changetype(T, S::SparseMatrixCLIL)
```
Comment (YingboMa) [inline src/systems/sparsematrixclil.jl:38]: `convert`?

Comment (YingboMa) [reply src/systems/sparsematrixclil.jl:38]: Or `SparseMatrixCLIL{T}`

File: src/systems/sparsematrixclil.jl [hunk 3]
```diff
@@ -189,13 +199,14 @@ function bareiss_update_virtual_colswap_mtk!(zero!, M::SparseMatrixCLIL, k, swap
             v == vpivot && continue
             ck = getcoeff(kvars, kcoeffs, v)
             ci = getcoeff(ivars, icoeffs, v)
-            ci = (pivot * ci - coeff * ck) ÷ last_pivot
-            if ci !== 0
+            p1 = Base.Checked.checked_mul(pivot, ci)
+            p2 = Base.Checked.checked_mul(coeff, ck)
+            ci = exactdiv(p1 - p2, last_pivot)
+            if ci != 0
```
Comment (YingboMa) [inline src/systems/sparsematrixclil.jl:205]:
```suggestion
            if !iszero(ci)
```

**Comment by YingboMa** [comment]
> Is it obvious that it always suffices to check the remainder or should we instead be checking for overflow on the multiplications?

Yes, we should check all integer ops.

File: src/systems/alias_elimination.jl (hunk 1 above)
Comment (oscardssmith) [reply src/systems/alias_elimination.jl:999]: I don't, but if I didn't convert it was complaining.

Comment (YingboMa) [reply src/systems/alias_elimination.jl:999]: See the type assert above

Comment (oscardssmith) [reply src/systems/alias_elimination.jl:999]: should I just remove the type assert?

Comment (YingboMa) [reply src/systems/alias_elimination.jl:999]: Maybe? But could you write the correct type assert according to the eltype?

File: src/systems/sparsematrixclil.jl [hunk 4]
```diff
@@ -189,13 +199,14 @@ function bareiss_update_virtual_colswap_mtk!(zero!, M::SparseMatrixCLIL, k, swap
             v == vpivot && continue
             ck = getcoeff(kvars, kcoeffs, v)
             ci = getcoeff(ivars, icoeffs, v)
-            ci = (pivot * ci - coeff * ck) ÷ last_pivot
-            if ci !== 0
+            p1 = Base.Checked.checked_mul(pivot, ci)
+            p2 = Base.Checked.checked_mul(coeff, ck)
+            ci = exactdiv(p1 - p2, last_pivot)
+            if !iszero(ci)
```
Comment (pepijndevos) [inline src/systems/sparsematrixclil.jl:205]: why not checked_subtract and checked_div rather than exactdiv?

Comment (oscardssmith) [reply src/systems/sparsematrixclil.jl:205]: div can't overflow, and checking exactdiv is a check that the subtraction doesn't overflow.

Comment (YingboMa) [reply src/systems/sparsematrixclil.jl:205]: > checking exactdiv is a check that the subtraction doesn't overflow.

I am not sure if that's strong enough though. We probably want to use `checked_subtract`.

Comment (Keno) [reply src/systems/sparsematrixclil.jl:205]: Yes, the subtraction should be checked. `exactdiv` is an algorithmic assertion though. If that's wrong, that error should propagate.

File: src/systems/alias_elimination.jl [hunk 5]
```diff
@@ -554,7 +554,15 @@ function aag_bareiss!(graph, var_to_diff, mm_orig::SparseMatrixCLIL)
     end
     solvable_variables = findall(is_linear_variables)
 
-    return mm, solvable_variables, do_bareiss!(mm, mm_orig, is_linear_variables)
+    local bar
+    try
+        bar = do_bareiss!(mm, mm_orig, is_linear_variables)
+    catch
```
Comment (Keno) [inline src/systems/alias_elimination.jl:560]: This should rethrow anything that's not an `OverflowError`.

File: src/structural_transformation/bareiss.jl [hunk 6]
```diff
@@ -124,8 +124,8 @@ function bareiss_update!(zero!, M::StridedMatrix, k, swapto, pivot,
             M[j, i], r = divrem(M[j, i] * pivot - M[j, k] * Mki, prev_pivot)
             flag = flag | r
         end
+        iszero(flag) || error("Overflow occurred")
```
Comment (Keno) [inline src/structural_transformation/bareiss.jl:127]: why this change?

File: src/systems/alias_elimination.jl [hunk 7]
```diff
@@ -240,8 +240,8 @@ end
     for i in range
         row = @view M[i, :]
         if constraint(count(!iszero, row))
+            iszero(val) && continue
```
Comment (Keno) [inline src/systems/alias_elimination.jl:243]: val is not defined here

File: src/structural_transformation/bareiss.jl (hunk 6 above)
Comment (oscardssmith) [reply src/structural_transformation/bareiss.jl:127]: it allows for early exit on overflow while keeping vectorization

Comment (Keno) [reply src/structural_transformation/bareiss.jl:127]: We should assume that overflow is unlikely. With the check outside the loop, the compiler has more opportunities for pipelining. In any case, ideally these kinds of changes need to be their own commit rather than being lumped in with a bugfix.

File: src/systems/alias_elimination.jl (hunk 1 above)
Comment (oscardssmith) [reply src/systems/alias_elimination.jl:999]: the problem here is that `AliasGraph` is currently an `AbstractDict{Int, Pair{Int, Int}}` and changing that to  `AbstractDict{Int, Pair{BigInt, Int}}` seems like a pretty dramatic change.

Comment (Keno) [reply src/systems/alias_elimination.jl:999]: We only support+-1 aliases at the moment anyway, so this should be fine

**Comment by oscardssmith** [comment]
Ok. I think this is ready to merge.

**Comment by oscardssmith** [comment]
I don't understand why format is failing here. JuliaFormatting isn't changing it.

**Comment by Keno** [comment]
format-check fails on master with the same error. I assume it's a JuliaFormatter version difference.

**Review by Keno — APPROVED** [review]

**Comment by YingboMa** [comment]
> format-check fails on master with the same error. I assume it's a JuliaFormatter version difference.

It's because the formatting on master is broken because of https://github.com/SciML/ModelingToolkit.jl/commit/a5f4e1e6abf941a6b2c2885228a8c659cef3094a @ChrisRackauckas
